From b618b3bdf41b57f135cda13e6139f4aaadb42072 Mon Sep 17 00:00:00 2001
From: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
Date: Mon, 25 Aug 2014 15:41:04 +0100
Subject: [PATCH 316/351] switch: arizona: Factor out headphone impedance
 setting

This allows custom jack detection code to specify the headphone
impedance since it is required by some of the arizona core code.

Change-Id: I51812e61c31bdf363566294ccf082ccf1edb4bc1
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
---
 drivers/switch/switch-arizona.c |   54 +++++++++++++++++----------------------
 include/linux/switch-arizona.h  |    3 +++
 2 files changed, 27 insertions(+), 30 deletions(-)

diff --git a/drivers/switch/switch-arizona.c b/drivers/switch/switch-arizona.c
index 0868576..6e9edcd 100644
--- a/drivers/switch/switch-arizona.c
+++ b/drivers/switch/switch-arizona.c
@@ -787,6 +787,28 @@ int arizona_wm1814_tune_headphone(struct arizona_extcon_info *info,
 	return 0;
 }
 
+void arizona_set_headphone_imp(struct arizona_extcon_info *info, int imp)
+{
+	struct arizona *arizona = info->arizona;
+
+	arizona->hp_impedance = imp;
+
+	if (arizona->pdata.hpdet_cb)
+		arizona->pdata.hpdet_cb(arizona->hp_impedance);
+
+	switch (arizona->type) {
+	case WM5110:
+		arizona_wm5110_tune_headphone(info, arizona->hp_impedance);
+		break;
+	case WM1814:
+		arizona_wm1814_tune_headphone(info, arizona->hp_impedance);
+		break;
+	default:
+		break;
+	}
+}
+EXPORT_SYMBOL_GPL(arizona_set_headphone_imp);
+
 int arizona_hpdet_start(struct arizona_extcon_info *info)
 {
 	struct arizona *arizona = info->arizona;
@@ -798,21 +820,7 @@ int arizona_hpdet_start(struct arizona_extcon_info *info)
 	if (info->arizona->pdata.fixed_hpdet_imp) {
 		int imp = info->arizona->pdata.fixed_hpdet_imp;
 
-		if (arizona->pdata.hpdet_cb)
-			arizona->pdata.hpdet_cb(imp);
-
-		switch (arizona->type) {
-		case WM5110:
-			arizona_wm5110_tune_headphone(info, imp);
-			info->arizona->hp_impedance = imp;
-			break;
-		case WM1814:
-			arizona_wm1814_tune_headphone(info, imp);
-			info->arizona->hp_impedance = imp;
-			break;
-		default:
-			break;
-		}
+		arizona_set_headphone_imp(info, imp);
 
 		ret = -EEXIST;
 		goto skip;
@@ -927,21 +935,7 @@ int arizona_hpdet_reading(struct arizona_extcon_info *info, int val)
 	if (val < 0)
 		return val;
 
-	arizona->hp_impedance = val;
-
-	if (arizona->pdata.hpdet_cb)
-		arizona->pdata.hpdet_cb(arizona->hp_impedance);
-
-	switch (arizona->type) {
-	case WM5110:
-		arizona_wm5110_tune_headphone(info, arizona->hp_impedance);
-		break;
-	case WM1814:
-		arizona_wm1814_tune_headphone(info, arizona->hp_impedance);
-		break;
-	default:
-		break;
-	}
+	arizona_set_headphone_imp(info, val);
 
 	if (info->mic) {
 		arizona_extcon_report(info, BIT_HEADSET);
diff --git a/include/linux/switch-arizona.h b/include/linux/switch-arizona.h
index c277c14..45f84c4d 100644
--- a/include/linux/switch-arizona.h
+++ b/include/linux/switch-arizona.h
@@ -53,6 +53,9 @@ struct arizona_jd_state {
 int arizona_jds_set_state(struct arizona_extcon_info *info,
 			  const struct arizona_jd_state *new_state);
 
+extern void arizona_set_headphone_imp(struct arizona_extcon_info *info,
+				      int imp);
+
 extern const struct arizona_jd_state arizona_hpdet_left;
 extern const struct arizona_jd_state arizona_hpdet_right;
 extern const struct arizona_jd_state arizona_micd_button;
-- 
1.7.9.5

